app/controllers/concerns/authorization.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+module Authorization
+  extend ActiveSupport::Concern
+
+  include ActionController::HttpAuthentication::Token
+
+  included do
+    before_action :set_token_and_options, :validate_token_and_options
+    before_action :set_access_token, :authenticate_access_token
+    before_action :authorize_access_token_as_readable, only: %i[index show]
+    before_action :authorize_access_token_as_writable, only: %i[create update destroy]
+  end
+
+  private
+
+  def token_string
+    @token_string ||= nil
+  end
+
+  def token_options
+    @token_options ||= nil
+  end
+
+  # NOTE: #token_and_options の仕様上、Authorizationヘッダのトークンタイプが Bearer だけでなく Token の場合も正当なフォーマットとして受理される。
+  # ここではどうしてもどちらか一方のみを有効にしなければならないわけではないので、そのままにしておく。
+  def set_token_and_options
+    @token_string, @token_options = token_and_options request
+  end
+
+  def validate_token_and_options
+    render problem: { error: 'invalid_request' }, status: :bad_request if token_string.blank? || token_options.present?
+  end
+
+  def access_token
+    @access_token ||= nil
+  end
+
+  def set_access_token
+    @access_token = AccessToken.find_by token: token_string
+  end
```

Comment:
ここも同じく attr_reader でよさそう。

もしくはこういう風に実装して、初回アクセス時に find_by するというのもよくやります。

```suggestion
  def access_token
    @access_token ||= AccessToken.find_by token: token_string
  end
```

Reply:
提案していただいた方法を利用するように変更しました。